File: packages/x/src/services/x-pro-deck.ts
// SPDX-License-Identifier: Apache-2.0
import type { XClient } from "../core/client";
import { GRYPHON_GRAPHQL_OPS } from "../config/x-pro-surface";

export interface XProDeckColumn {
  rest_id: string;
  pathname: string;
  width?: string;
  media_preview?: string;
  latest?: boolean;
  hide_header?: boolean;
}

export interface XProDeckConfig {
  title?: string;
  icon?: string;
  is_pinned?: boolean;
}

export interface XProDeck {
  rest_id: string;
  config?: XProDeckConfig;
  deck_columns_v2?: XProDeckColumn[];
}

export interface XProAccountSync {
  active_deck_id?: string;
  composer_expanded?: boolean;
  default_column_width?: string;
  default_media_preview?: string;
  navbar_expanded?: boolean;
}

export interface ViewerAccountSyncResult {
  decks: XProDeck[];
  accountsync_client_config?: XProAccountSync;
  accountsync_onboarding_state?: Record<string, unknown>;
  raw: unknown;
}

function dig<T = unknown>(obj: unknown, path: string[]): T | undefined {
  let cur: unknown = obj;
  for (const key of path) {
    if (cur == null || typeof cur !== "object") return undefined;
    cur = (cur as Record<string, unknown>)[key];
  }
  return cur as T | undefined;
}

export function parseViewerAccountSync(json: unknown): ViewerAccountSyncResult {
  const viewer = dig<Record<string, unknown>>(json, ["data", "viewer_v2"]);
  const decks = (viewer?.decks as XProDeck[] | undefined) ?? [];
  return {
    decks,
    accountsync_client_config: viewer?.accountsync_client_config as
      | XProAccountSync
      | undefined,
    accountsync_onboarding_state: dig(json, [
      "data",
      "accountsync_onboarding_state",
    ]) as Record<string, unknown> | undefined,
    raw: json,
  };
}

export async function viewerAccountSync(
  client: XClient,
): Promise<ViewerAccountSyncResult> {
  const raw = await client.graphqlWaiting("ViewerAccountSync", {});
  return parseViewerAccountSync(raw);
}

export async function getDeck(
  client: XClient,
  deckId: string,
): Promise<XProDeck | null> {
  const sync = await viewerAccountSync(client);
  return sync.decks.find((d) => d.rest_id === deckId) ?? null;
}

export async function createDeck(
  client: XClient,
  name: string,
  columns: Array<{ pathname: string; width?: string }> = [],
): Promise<string | null> {
  const raw = await client.graphqlWaiting("CreateDeck", { name, columns });
  return dig<string>(raw, ["data", "deck_insert", "rest_id"]) ?? null;
}

export async function updateDeck(
  client: XClient,
  deckId: string,
  config: XProDeckConfig,
): Promise<unknown> {
  return client.graphqlWaiting("UpdateDeck", { deckId, config });
}

export async function removeDeck(
  client: XClient,
  deckId: string,
): Promise<unknown> {
  return client.graphqlWaiting("RemoveDeck", { deckId });
}

export async function reorderDecks(
  client: XClient,
  deckIds: string[],
): Promise<unknown> {
  return client.graphqlWaiting("ReorderDecks", { deckIds });
}

export async function createColumn(
  client: XClient,
  deckId: string,
  column: { pathname: string; width?: string; media_preview?: string },
): Promise<unknown> {
  return client.graphqlWaiting("CreateColumn", { deckId, column });
}

export async function updateColumn(
  client: XClient,
  deckId: string,
  columnId: string,
  column: Partial<XProDeckColumn>,
): Promise<unknown> {
  return client.graphqlWaiting("UpdateColumn", { deckId, columnId, column });
}

export async function removeColumn(
  client: XClient,
  deckId: string,
  columnId: string,
): Promise<unknown> {
  return client.graphqlWaiting("RemoveColumn", { deckId, columnId });
}

export async function reorderColumns(
  client: XClient,
  deckId: string,
  columnIds: string[],
): Promise<unknown> {
  return client.graphqlWaiting("ReorderColumns", { deckId, columnIds });
}

export async function importClientSyncColumns(client: XClient): Promise<unknown> {
  return client.graphqlWaiting("GryphonImportClientSyncColumns", {});
}

export async function probeXProAccess(client: XClient): Promise<{
  ok: boolean;
  deck_count: number;
  active_deck_id?: string;
  graphql_ops: readonly string[];
  error?: string;
}> {
  try {
    const sync = await viewerAccountSync(client);
    return {
      ok: true,
      deck_count: sync.decks.length,
      active_deck_id: sync.accountsync_client_config?.active_deck_id,
      graphql_ops: GRYPHON_GRAPHQL_OPS,
    };
  } catch (err: unknown) {
    return {
      ok: false,
      deck_count: 0,
      graphql_ops: GRYPHON_GRAPHQL_OPS,
      error: err instanceof Error ? err.message : String(err),
    };
  }
}

File: rust-bridge/crates/x-client/src/x_pro_deck.rs
// SPDX-License-Identifier: Apache-2.0
//! X Pro (Gryphon) deck GraphQL — ViewerAccountSync, CreateDeck, columns.

use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::client::XClient;
use crate::Result;

pub use crate::surface::{GRYPHON_GRAPHQL_OPS, X_PRO_RECON_URLS};

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct XProDeckConfig {
    #[serde(default)]
    pub title: Option<String>,
    #[serde(default)]
    pub icon: Option<String>,
    #[serde(default)]
    pub is_pinned: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct XProDeckColumn {
    pub rest_id: String,
    pub pathname: String,
    #[serde(default)]
    pub width: Option<String>,
    #[serde(default)]
    pub media_preview: Option<String>,
    #[serde(default)]
    pub latest: Option<bool>,
    #[serde(default)]
    pub hide_header: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct XProDeck {
    pub rest_id: String,
    #[serde(default)]
    pub config: Option<XProDeckConfig>,
    #[serde(default)]
    pub deck_columns_v2: Vec<XProDeckColumn>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct XProAccountSyncConfig {
    #[serde(default)]
    pub active_deck_id: Option<String>,
    #[serde(default)]
    pub composer_expanded: Option<bool>,
    #[serde(default)]
    pub default_column_width: Option<String>,
    #[serde(default)]
    pub default_media_preview: Option<String>,
    #[serde(default)]
    pub navbar_expanded: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ViewerAccountSyncResult {
    pub decks: Vec<XProDeck>,
    #[serde(default)]
    pub accountsync_client_config: Option<XProAccountSyncConfig>,
    #[serde(default)]
    pub accountsync_onboarding_state: Option<Value>,
    pub raw: Value,
}

fn dig<'a>(v: &'a Value, path: &[&str]) -> Option<&'a Value> {
    let mut cur = v;
    for key in path {
        cur = cur.get(key)?;
    }
    Some(cur)
}

pub fn parse_viewer_account_sync(json: &Value) -> ViewerAccountSyncResult {
    let viewer = dig(json, &["data", "viewer_v2"]);
    let decks: Vec<XProDeck> = viewer
        .and_then(|v| v.get("decks"))
        .and_then(|d| serde_json::from_value(d.clone()).ok())
        .unwrap_or_default();
    let accountsync_client_config = viewer
        .and_then(|v| v.get("accountsync_client_config"))
        .and_then(|c| serde_json::from_value(c.clone()).ok());
    let accountsync_onboarding_state =
        dig(json, &["data", "accountsync_onboarding_state"]).cloned();
    ViewerAccountSyncResult {
        decks,
        accountsync_client_config,
        accountsync_onboarding_state,
        raw: json.clone(),
    }
}

impl XClient {
    /// `ViewerAccountSync` — list decks + client config (X Pro).
    pub async fn viewer_account_sync(&self) -> Result<ViewerAccountSyncResult> {
        let json = self.graphql("ViewerAccountSync", serde_json::json!({}), None).await?;
        Ok(parse_viewer_account_sync(&json))
    }

    pub async fn xpro_get_deck(&self, deck_id: &str) -> Result<Option<XProDeck>> {
        let sync = self.viewer_account_sync().await?;
        Ok(sync.decks.into_iter().find(|d| d.rest_id == deck_id))
    }

    pub async fn xpro_create_deck(
        &self,
        name: &str,
        columns: Value,
    ) -> Result<Option<String>> {
        let json = self
            .graphql(
                "CreateDeck",
                serde_json::json!({ "name": name, "columns": columns }),
                None,
            )
            .await?;
        Ok(dig(&json, &["data", "deck_insert", "rest_id"])
            .and_then(|v| v.as_str())
            .map(str::to_owned))
    }

    pub async fn xpro_remove_deck(&self, deck_id: &str) -> Result<Value> {
        self.graphql("RemoveDeck", serde_json::json!({ "deckId": deck_id }), None)
            .await
    }

    pub async fn xpro_update_deck(
        &self,
        deck_id: &str,
        config: &XProDeckConfig,
    ) -> Result<Value> {
        self.graphql(
            "UpdateDeck",
            serde_json::json!({ "deckId": deck_id, "config": config }),
            None,
        )
        .await
    }

    pub async fn probe_xpro_access(&self) -> Result<(bool, usize, Option<String>)> {
        match self.viewer_account_sync().await {
            Ok(sync) => Ok((
                true,
                sync.decks.len(),
                sync.accountsync_client_config
                    .and_then(|c| c.active_deck_id),
            )),
            Err(e) => Err(e),
        }
    }
}